src/DynamoUtilities/DynamoLock.cs

```diff
@@ -0,0 +1,79 @@
+
+        internal DynamoLock()
+        {
+            rwlock = new ReaderWriterLockSlim();
+        }
+
+        /// <summary>
+        /// Locks for read. Use this API when you need more control of the lock lifetime
+        /// </summary>
+        internal void LockForRead() => rwlock.EnterReadLock();
+        /// <summary>
+        /// Locks for write. Use this API when you need more control of the lock lifetime
+        /// </summary>
+        internal void LockForWrite() => rwlock.EnterWriteLock();
+        /// <summary>
+        /// Locks for upgradeable read. Use this API when you need more control of the lock lifetime
+        /// </summary>
+        internal void LockForUpgradeableRead() => rwlock.EnterUpgradeableReadLock();
+
+        /// <summary>
+        /// Unlocks a read lock. Use this API if you called LockForRead beforehand.
+        /// </summary>
+        internal void UnlockForRead() => rwlock.ExitReadLock();
+        /// <summary>
+        /// Unlocks a write lock. Use this API if you called LockForWrite beforehand.
+        /// </summary>
+        internal void UnlockForWrite() => rwlock.ExitWriteLock();
+        /// <summary>
+        /// Unlocks an upgradeable read lock. Use this API if you called LockForUpgradeableRead beforehand.
+        /// </summary>
+        internal void UnlockForUpgradeableRead() => rwlock.ExitUpgradeableReadLock();
+
+
+        public void Dispose()
+        {
+            try
+            {
+                if (rwlock.IsReadLockHeld)
+                    rwlock.ExitReadLock();
+                if (rwlock.IsWriteLockHeld)
+                    rwlock.ExitWriteLock();
+                if (rwlock.IsUpgradeableReadLockHeld)
+                    rwlock.ExitUpgradeableReadLock();
+            }
+            finally 
+            {
+                rwlock.Dispose();
+            }
+        }
+
+        /// <summary>
+        /// Constructs a disposable object that locks for write and releases the lock when Dispose is called.
+        /// </summary>
+        /// <returns>New disposable object</returns>
+        internal IDisposable CreateWriteLock() => Disposable.Create(() => { rwlock.EnterWriteLock(); }, () => { rwlock.ExitWriteLock(); });
+
+        /// <summary>
+        /// Constructs a disposable object that locks for read and releases the lock when Dispose is called.
+        /// </summary>
+        /// <returns>New disposable object</returns>
+        internal IDisposable CreateReadLock() => Disposable.Create(() => { rwlock.EnterReadLock(); }, () => { rwlock.ExitReadLock(); });
+
+        /// <summary>
+        /// Constructs a disposable object that locks for read and allows lock for write in the same scope. Releases the lock when Dispose is called.
+        /// </summary>
+        /// <returns>New disposable object</returns>
+        internal IDisposable CreateUpgradeableReadLock() => Disposable.Create(() => { rwlock.EnterUpgradeableReadLock(); }, () => { rwlock.ExitUpgradeableReadLock(); });
```

Comment:
It doesn't look like this is being used, do we need it, and what is its purpose?

Reply:
Not used now. But it might be useful. It exposes the remaining lock style.
It is useful when you have a code scope that does read and write. Using upgradeable is a faster and cleaner solution
Ex
```
getOrCreateStuff() {
   using(dynLock.CreateUpgradeableLock())
   {
       if (arr.HasItem()) {
           return arr.Item;
       }
       using(dynLock.CreateWriteableLock())
       {
          return arr.CreateItem();
       }
   }
}
```

Reply:
Because `ReaderWriterLockSlim` does not have one.

Reply:
I meant to ask why `ReaderWriterLockSlim` doesn't have one but that's okay.